internal/bft/controller.go

```diff
@@ -592,104 +592,93 @@ func (c *Controller) sync() (viewNum uint64, seq uint64, decisions uint64) {
 		c.ViewChanger.close()
 	}
 
-	decision := syncResponse.Latest
-	if decision.Proposal.Metadata == nil {
-		c.Logger.Infof("Synchronizer returned with proposal metadata nil")
-		response := c.fetchState()
-		if response == nil {
-			return 0, 0, 0
+	// The synchronizer returns a response which includes the latest decision with its proposal metadata.
+	// This proposal may be empty (its metadata is empty), meaning the synchronizer is not aware of any decisions made.
+	// Otherwise, the latest proposal sequence returned may be higher than our latest sequence, meaning we should
+	// update the checkpoint.
+	// In other cases we should not update the checkpoint.
+	// However, we always must fetch the latest state from other nodes,
+	// since the view may have advanced without this node and with no decisions.
+
+	var viewNumToReturn, seqToReturn, decisionsToReturn uint64
+
+	latestDecision := syncResponse.Latest
+	var latestDecisionSeq, latestDecisionViewNum, latestDecisionDecisions uint64
+	var latestDecisionMetadata *protos.ViewMetadata
+	if len(latestDecision.Proposal.Metadata) == 0 {
+		c.Logger.Infof("Synchronizer returned with an empty proposal metadata")
+		latestDecisionMetadata = nil
+	} else {
+		md := &protos.ViewMetadata{}
+		if err := proto.Unmarshal(latestDecision.Proposal.Metadata, md); err != nil {
+			c.Logger.Panicf("Controller was unable to unmarshal the proposal metadata returned by the Synchronizer")
 		}
-		if response.View > 0 && response.Seq == 1 {
-			c.Logger.Infof("The collected state is with view %d and sequence %d", response.View, response.Seq)
-			newViewToSave := &protos.SavedMessage{
-				Content: &protos.SavedMessage_NewView{
-					NewView: &protos.ViewMetadata{
-						ViewId:          response.View,
-						LatestSequence:  0,
-						DecisionsInView: 0,
-					},
-				},
-			}
-			if err := c.State.Save(newViewToSave); err != nil {
-				c.Logger.Panicf("Failed to save message to state, error: %v", err)
-			}
-			c.ViewChanger.InformNewView(response.View)
-			return response.View, 1, 0
-		}
-		return 0, 0, 0
+		latestDecisionSeq = md.LatestSequence
+		latestDecisionViewNum = md.ViewId
+		latestDecisionDecisions = md.DecisionsInView
+		latestDecisionMetadata = md
 	}
-	md := &protos.ViewMetadata{}
-	if err := proto.Unmarshal(decision.Proposal.Metadata, md); err != nil {
-		c.Logger.Panicf("Controller was unable to unmarshal the proposal metadata returned by the Synchronizer")
+
+	controllerSequence := c.latestSeq()
+	seqToReturn = controllerSequence + 1
```

Comment:
I know this is a nit, but, can we call `seqToReturn`: `newProposalSequence` or something? Because it's really hard to understand the increment otherwise without understanding the semantics of the sequence we return.